patches/api/0332-Better-Stats-API.patch

```diff
@@ -0,0 +1,721 @@
++
++    private static final Map<NamespacedKey, StatisticType<?>> STATISTIC_TYPE_MAP = Maps.newHashMap();
++
++    @Nullable
++    public static StatisticType<?> getStatisticType(@NotNull NamespacedKey key) {
++        return STATISTIC_TYPE_MAP.get(key);
++    }
++
++    @NotNull
++    public static Collection<StatisticType<?>> values() {
++        return STATISTIC_TYPE_MAP.values();
++    }
++
++    private static final Predicate<EntityType> NOT_UNKNOWN_CHECK = entityType -> entityType != EntityType.UNKNOWN;
++
++    public static final StatisticType<Material> BLOCK_MINED = new StatisticType<>("mined", Material::isBlock);
++    public static final StatisticType<Material> ITEM_CRAFTED = new StatisticType<>("crafted", Material::isItem);
++    public static final StatisticType<Material> ITEM_USED = new StatisticType<>("used", Material::isItem);
++    public static final StatisticType<Material> ITEM_BROKEN = new StatisticType<>("broken", Material::isItem);
++    public static final StatisticType<Material> ITEM_PICKED_UP = new StatisticType<>("picked_up", Material::isItem);
++    public static final StatisticType<Material> ITEM_DROPPED = new StatisticType<>("dropped", Material::isItem);
++    public static final StatisticType<EntityType> ENTITY_KILLED = new StatisticType<>("killed", NOT_UNKNOWN_CHECK);
++    public static final StatisticType<EntityType> ENTITY_KILLED_BY = new StatisticType<>("killed_by", NOT_UNKNOWN_CHECK);
++    public static final StatisticType<CustomStatistic> CUSTOM_STATS = new StatisticType<>("custom");
++
++    private final NamespacedKey key;
++    private final Map<B, Statistic<B>> statisticMap = new IdentityHashMap<>();
++    private final Predicate<B> typeCheck;
++
++    private StatisticType(String key) {
++        this(key, Predicates.alwaysTrue());
++    }
++
++    private StatisticType(String key, Predicate<B> typeCheck) {
++        this.key = NamespacedKey.minecraft(key);
++        this.typeCheck = typeCheck;
++        STATISTIC_TYPE_MAP.put(this.key, this);
++    }
++
++    /**
++     * Creates or gets the statistic from this type for the specified "thing".
++     *
++     * @param thing what you want the stat of
++     * @return the statistic for that thing
++     * @throws IllegalArgumentException if the thing is not valid for this {@link StatisticType}
++     */
++    @NotNull
++    public Statistic<B> of(@NotNull B thing) {
++        if (!typeCheck.test(thing)) {
++            throw new IllegalArgumentException(thing + " is not valid for " + this.getKey());
++        }
++        if (this == StatisticType.CUSTOM_STATS) {
++            return this.statisticMap.get(thing);
++        }
++        return this.statisticMap.computeIfAbsent(thing, newValue -> new Statistic<>(thing, this));
++    }
++
++    @NotNull
++    B registerCustomStatistic(@NotNull B stat) {
++        if (this != StatisticType.CUSTOM_STATS) {
++            throw new IllegalArgumentException("Must be the CUSTOM_STATS stat type");
++        }
```

Comment:
_Cries in NotNull_
I also don't see any reason to make custom stats something special here. All stat types should be treated equally.

Reply:
because all the other stat types are created as-needed. the custom one is special. This isnt API anyways since its not public, but I put it there more for safety. API just uses the `of` method to get/create Statistic instances